sentinelsat/sentinel.py

```diff
@@ -502,99 +546,30 @@ def download(self, id, directory_path=".", checksum=True, **kwargs):
-        path = Path(directory_path) / filename
-        product_info["path"] = str(path)
-        product_info["downloaded_bytes"] = 0
-
-        self.logger.info("Downloading %s to %s", id, path)
-
-        if path.exists():
-            # We assume that the product has been downloaded and is complete
-            return product_info
-
-        # An incomplete download triggers the retrieval from the LTA if the product is not online
-        if not self.is_online(id):
-            self.trigger_offline_retrieval(id)
-            raise LTATriggered(id)
-
-        self._download_outer(product_info, path, checksum)
-        return product_info
-
-    def _download_outer(self, product_info: Dict[str, Any], path: Path, verify_checksum: bool):
-        # Use a temporary file for downloading
-        temp_path = path.with_name(path.name + ".incomplete")
-        skip_download = False
-        if temp_path.exists():
-            size = temp_path.stat().st_size
-            if size > product_info["size"]:
-                self.logger.warning(
-                    "Existing incomplete file %s is larger than the expected final size"
-                    " (%s vs %s bytes). Deleting it.",
-                    str(temp_path),
-                    size,
-                    product_info["size"],
-                )
-                temp_path.unlink()
-            elif size == product_info["size"]:
-                if verify_checksum and not self._checksum_compare(temp_path, product_info):
-                    # Log a warning since this should never happen
-                    self.logger.warning(
-                        "Existing incomplete file %s appears to be fully downloaded but "
-                        "its checksum is incorrect. Deleting it.",
-                        str(temp_path),
-                    )
-                    temp_path.unlink()
-                else:
-                    skip_download = True
-            else:
-                # continue downloading
-                self.logger.info(
-                    "Download will resume from existing incomplete file %s.", temp_path
-                )
-                pass
-        if not skip_download:
-            # Store the number of downloaded bytes for unit tests
-            temp_path.parent.mkdir(parents=True, exist_ok=True)
-            product_info["downloaded_bytes"] = self._download(
-                product_info["url"], temp_path, product_info["size"]
-            )
-        # Check integrity with MD5 checksum
-        if verify_checksum is True:
-            if not self._checksum_compare(temp_path, product_info):
-                temp_path.unlink()
-                raise InvalidChecksumError("File corrupt: checksums do not match")
-        # Download successful, rename the temporary file to its proper name
-        shutil.move(temp_path, path)
+        downloader = Downloader(self, node_filter=nodefilter, verify_checksum=checksum)
```

Comment:
That's a good suggestion and I considered doing it that way, but it would cause issues if the download-related functions were called in parallel, since the configuration params are shared. A reasonable compromise might be to make it a member of `SentinelAPI` and `copy()` the downloader instance each time it is used, so the config can be safely modified.

Reply:
Or that you pass in the class during `__init__` but it gets instantiated in `SentinelAPI.download`?

Reply:
I went with the `copy()` approach.